MM-23484 - add simple validation to getUserStatusesByIds

TheoVitkovskiy opened the pull request:
**Summary**

Adds extra validation (just a simple length check) for the user_ids that are passed to the `/status/ids` endpoint.

For context, here is the link to the discussion:
https://community.mattermost.com/core/pl/16ei5i6c8bfrug37nt4pgewute

**Ticket Link**

https://mattermost.atlassian.net/browse/MM-23484

streamer45 commented on api4/status_test.go:
```diff
@@ -92,6 +92,14 @@ func TestGetUsersStatusesByIds(t *testing.T) {
 
+	invalidUserIds := []string{"invalid_user_id", "invalid_user_id"}
+	_, resp = Client.GetUsersStatusesByIds(invalidUserIds)
+	CheckBadRequestStatus(t, resp)
```
I think we can also add a case where the input is an empty slice.

streamer45 commented on api4/status.go:
```diff
@@ -40,7 +40,19 @@ func getUserStatus(c *Context, w http.ResponseWriter, r *http.Request) {
+
+	for _, userId := range userIds {
+		if len(userId) != 26 {
+			isUserIdsInvalid = true
```
I think we can simplify and return directly here. No need for the additional variable. We can stop and return as soon as we find something invalid.

TheoVitkovskiy replied:
my only reason was to avoid code duplication and improve readability a bit. Do you want to avoid the variable, because this function is often called?
Or should I treat it as a general rule for future code changes?

streamer45 replied:
My reasoning is to return early to keep the whole logic easier to understand. 
If we get an empty slice there's no need to go any further. It's an unrecoverable error.

Avoiding code duplication is great but in this specific case I don't think we are gaining much on that front.

streamer45 commented on api4/status.go:
```diff
@@ -40,7 +40,19 @@ func getUserStatus(c *Context, w http.ResponseWriter, r *http.Request) {
+	isUserIdsInvalid := false
+
 	if len(userIds) == 0 {
+		isUserIdsInvalid = true
```
Same here, we can just set the error and return.

TheoVitkovskiy commented:
> Thank you for your contribution @TheoVitkovskiy !
> I've left a couple of suggestions but overall looks good +1

@streamer45 I addressed your points, and added the changes in an extra commit.

streamer45 commented on api4/status_test.go:
```diff
@@ -92,6 +92,18 @@ func TestGetUsersStatusesByIds(t *testing.T) {
+
+	invalidUserIds = []string{"", ""}
+	_, resp = Client.GetUsersStatusesByIds(invalidUserIds)
+	CheckBadRequestStatus(t, resp)
```
This is nice to have too but what I actually meant (sorry for not being super clear) was an empty slice `[]string{}` so we can test the first return case.

TheoVitkovskiy replied:
was a bit confused by slice, added it

streamer45 replied:
No problem. That's the name Golang [uses](https://blog.golang.org/slices-intro) for that array-like data structure.

jespino commented on api4/status_test.go:
```diff
@@ -92,6 +92,18 @@ func TestGetUsersStatusesByIds(t *testing.T) {
+	_, resp = Client.GetUsersStatusesByIds(invalidUserIds)
+	CheckBadRequestStatus(t, resp)
+
+	invalidUserIds = []string{"", ""}
```
Would be great to have another test with an empty list of userIds.

TheoVitkovskiy replied:
added it, thank you for the review!

jespino replied:
That was fast! Thanks!

TheoVitkovskiy commented:
> Looks good! Thank you again @TheoVitkovskiy +1

No problem! It was a very good introduction to the project, had a lot of fun doing it.

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-server-pr-14158.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mattermod commented:
Test server destroyed

streamer45 commented:
@TheoVitkovskiy could you help resolving the conflicts so we can merge this? Thanks!

TheoVitkovskiy commented:
> @TheoVitkovskiy could you help resolving the conflicts so we can merge this? Thanks!

@streamer45 should I rebase against master and push force? Or do I merge master into my feature branch?

streamer45 commented:
> > @TheoVitkovskiy could you help resolving the conflicts so we can merge this? Thanks!
> 
> @streamer45 should I rebase against master and push force? Or do I merge master into my feature branch?

Since the PR has already been reviewed already, we recommend to merge master into your feature branch, fix the remaining conflicts, and do a normal push.

TheoVitkovskiy commented:
> > > @TheoVitkovskiy could you help resolving the conflicts so we can merge this? Thanks!
> > 
> > 
> > @streamer45 should I rebase against master and push force? Or do I merge master into my feature branch?
> 
> Since the PR has been reviewed already, we recommend to merge master into your feature branch, fix the remaining conflicts, and do a normal push.

I did, the tests are much nicer now! 

> > > @TheoVitkovskiy could you help resolving the conflicts so we can merge this? Thanks!
> > 
> > 
> > @streamer45 should I rebase against master and push force? Or do I merge master into my feature branch?
> 
> Since the PR has been reviewed already, we recommend to merge master into your feature branch, fix the remaining conflicts, and do a normal push.

Fixed the merge conflicts. Thanks to @jespino the tests are much nicer now!
